cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java

```diff
@@ -275,57 +297,129 @@ public List<TopicId> listTopics(NamespaceId namespaceId)
+
+      // Check if adding these mutations will exceed the batch size limit.
+      // If exceeding limit, publish the current batch and start a new one.
+      // Ensure that all messages within a StoreRequest are published in a single batch to maintain atomicity.
+      // If the batch publish succeeds, all messages are persisted,
+      // and the corresponding future is completed successfully.
+      // If the batch publish fails, none of the messages are persisted,
+      // and the future is completed exceptionally, allowing the caller to retry the entire StoreRequest.
+      AtomicLong currentBatchSize = new AtomicLong(0);
+      currentRequest.forEach(payload -> currentBatchSize.addAndGet(payload.length));
+      if (estimatedBatchSize + currentBatchSize.get() > MAX_BATCH_SIZE_IN_BYTES) {
+        commitPublishMutations(mutations, publishFutures);
+        estimatedBatchSize = 0;
+        sequenceId.set(0);
+      }
 
-        if (batch.isEmpty() && (sequenceId < publishBatchSize
-            || System.currentTimeMillis() - start < publishBatchTimeoutMillis)) {
-          try {
-            Thread.sleep(publishDelayMillis);
-          } catch (InterruptedException e) {
-            throw new IOException(e);
+      List<Mutation> currentRequestMutations = generatePublishMutations(currentRequest, sequenceId);
+      mutations.addAll(currentRequestMutations);
+      publishFutures.add(requestFuture.getFuture());
+      estimatedBatchSize += currentBatchSize.get();
+
+      if (batch.isEmpty() && (sequenceId.get() < publishBatchSize
+          || System.currentTimeMillis() - start < publishBatchTimeoutMillis)) {
+        try {
+          Thread.sleep(publishDelayMillis);
+        } catch (InterruptedException e) {
+          for (CompletableFuture<Void> future : publishFutures) {
+            future.completeExceptionally(e);
           }
         }
       }
-      if (!batchCopy.isEmpty()) {
-        try {
-          client.write(batchCopy);
-        } catch (SpannerException e) {
-          throw new IOException(e);
+    }
+    commitPublishMutations(mutations, publishFutures);
+  }
+
+  private void commitPublishMutations(List<Mutation> mutations,
+      List<CompletableFuture<Void>> futures) {
+    if (!mutations.isEmpty()) {
+      try {
+        client.write(mutations);
+        LOG.trace("published mutations : {}", mutations.size());
+        for (CompletableFuture<Void> future : futures) {
+          future.complete(null);
         }
+      } catch (SpannerException e) {
+        for (CompletableFuture<Void> future : futures) {
+          future.completeExceptionally(e);
+        }
+      } finally {
+        mutations.clear();
+        futures.clear();
       }
     }
-    //TODO: Add a RollbackDetail implementation that throws exceptions if any of the methods is called.
-    return null;
+  }
+
+  private List<Mutation> generatePublishMutations(StoreRequest request, AtomicInteger sequenceId) {
```

Comment:
Why this is generatePublishMutations and the other one createPublishMutation?
I'd simplify these and rename them to createMutations and createMutation